p2p/http/libp2phttp.go

```diff
@@ -0,0 +1,754 @@
+	logging "github.com/ipfs/go-log/v2"
+	host "github.com/libp2p/go-libp2p/core/host"
+	"github.com/libp2p/go-libp2p/core/network"
+	"github.com/libp2p/go-libp2p/core/peer"
+	"github.com/libp2p/go-libp2p/core/protocol"
+	gostream "github.com/libp2p/go-libp2p/p2p/gostream"
+	ma "github.com/multiformats/go-multiaddr"
+)
+
+var log = logging.Logger("libp2phttp")
+
+const ProtocolIDForMultistreamSelect = "/http/1.1"
+const peerMetadataLimit = 8 << 10 // 8KB
+const peerMetadataLRUSize = 256   // How many different peer's metadata to keep in our LRU cache
+
+// TODOs:
+// - integrate with the conn gater and resource manager
+
+// ProtocolMeta is metadata about a protocol.
+type ProtocolMeta struct {
+	// Path defines the HTTP Path prefix used for this protocol
+	Path string `json:"path"`
+}
+
+type PeerMeta map[protocol.ID]ProtocolMeta
+
+// WellKnownHandler is an http.Handler that serves the .well-known/libp2p resource
+type WellKnownHandler struct {
+	wellknownMapMu   sync.Mutex
+	wellKnownMapping PeerMeta
+}
+
+// streamHostListen retuns a net.Listener that listens on libp2p streams for HTTP/1.1 messages.
+func streamHostListen(streamHost host.Host) (net.Listener, error) {
+	return gostream.Listen(streamHost, ProtocolIDForMultistreamSelect)
+}
+
+func (h *WellKnownHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	// Check if the requests accepts JSON
+	accepts := r.Header.Get("Accept")
+	if accepts != "" && !(strings.Contains(accepts, "application/json") || strings.Contains(accepts, "*/*")) {
+		http.Error(w, "Only application/json is supported", http.StatusNotAcceptable)
+		return
+	}
+
+	if r.Method != http.MethodGet {
+		http.Error(w, "Only GET requests are supported", http.StatusMethodNotAllowed)
+		return
+	}
+
+	// Return a JSON object with the well-known protocols
+	h.wellknownMapMu.Lock()
+	mapping, err := json.Marshal(h.wellKnownMapping)
+	h.wellknownMapMu.Unlock()
+	if err != nil {
+		http.Error(w, "Marshal error", http.StatusInternalServerError)
+		return
+	}
+	w.Header().Add("Content-Type", "application/json")
+	w.Header().Add("Content-Length", strconv.Itoa(len(mapping)))
+	w.Write(mapping)
+}
+
+func (h *WellKnownHandler) AddProtocolMapping(p protocol.ID, protocolMeta ProtocolMeta) {
+	h.wellknownMapMu.Lock()
+	if h.wellKnownMapping == nil {
+		h.wellKnownMapping = make(map[protocol.ID]ProtocolMeta)
+	}
+	h.wellKnownMapping[p] = protocolMeta
+	h.wellknownMapMu.Unlock()
+}
+
+func (h *WellKnownHandler) RemoveProtocolMapping(p protocol.ID) {
+	h.wellknownMapMu.Lock()
+	if h.wellKnownMapping != nil {
+		delete(h.wellKnownMapping, p)
+	}
+	h.wellknownMapMu.Unlock()
+}
+
+// HTTPHost is a libp2p host for request/responses with HTTP semantics. This is
+// in contrast to a stream-oriented host like the host.Host interface. Its
+// zero-value (&HTTPHost{}) is usable. Do not copy by value.
+// See examples for usage.
+//
+//	Warning, this is experimental. The API will likely change.
+type HTTPHost struct {
```

Comment:
Should this just be called `Host`? Users will use this as `libp2phttp.Host` so it should be easy to differentiate against the core stream-based `host.Host` type.

Reply:
Yes. Avoids stutter. If everything were defined in a single package, then the type would be `Libp2pHTTPHost`, so `libp2phttp.Host` is appropriate.